python/ray/serve/batching.py

```diff
@@ -206,35 +215,43 @@ async def _process_batches(self, func: Callable) -> None:
         """Loops infinitely and processes queued request batches."""
 
         while True:
-            batch: List[_SingleRequest] = await self.wait_for_batch()
-            assert len(batch) > 0
-            self_arg = batch[0].self_arg
-            args, kwargs = _batch_args_kwargs([item.flattened_args for item in batch])
-            futures = [item.future for item in batch]
-
-            # Method call.
-            if self_arg is not None:
-                func_future_or_generator = func(self_arg, *args, **kwargs)
-            # Normal function call.
-            else:
-                func_future_or_generator = func(*args, **kwargs)
-
-            if isasyncgenfunction(func):
-                func_generator = func_future_or_generator
-                await self._consume_func_generator(func_generator, futures, len(batch))
-            else:
-                try:
-                    func_future = func_future_or_generator
-                    results = await func_future
-                    self._validate_results(results, len(batch))
-                    for result, future in zip(results, futures):
-                        # If the client has disconnected, the future is canceled.
-                        # We should only set result when the connection is still live.
-                        if not future.cancelled():
-                            future.set_result(result)
-                except Exception as e:
-                    for future in futures:
-                        future.set_exception(e)
+            try:
+                self.curr_iteration_start_time = time.time()
+                batch: List[_SingleRequest] = await self.wait_for_batch()
+                assert len(batch) > 0
+                self_arg = batch[0].self_arg
+                args, kwargs = _batch_args_kwargs([item.flattened_args for item in batch])
+                futures = [item.future for item in batch]
+
+                # Method call.
+                if self_arg is not None:
+                    func_future_or_generator = func(self_arg, *args, **kwargs)
+                # Normal function call.
+                else:
+                    func_future_or_generator = func(*args, **kwargs)
+
+                if isasyncgenfunction(func):
+                    func_generator = func_future_or_generator
+                    await self._consume_func_generator(func_generator, futures, len(batch))
```

Comment:
We also need to carefully refactor _consume_func_generator to guarantee that no future could be set twice

Reply:
I pulled the result/exception-setting logic into functions that guard against setting finished futures. That should be sufficient to prevent a future from being set twice.